### DISCUSSION TITLE
Why parent binding no working in my code?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#12389] [Q&A] [answered]
The version I am using is 11.0.1. The target framework I am using is .Net 7.

I created a TemplatedControl inherits from Button. I want to set its TextDecorations by code.

Here is my code:
```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:Test.Controls">
	<Design.PreviewWith>
		<controls:SButton />
	</Design.PreviewWith>
  <Style Selector="controls|SButton">

	  <Setter Property="Template">
		  <ControlTemplate>
			  <Border Name="B" ClipToBounds="True" Background="{TemplateBinding Background}" BorderBrush="{TemplateBinding BorderBrush}" BorderThickness="{TemplateBinding BorderThickness}"  CornerRadius="{TemplateBinding CornerRadius}">
				  <ContentPresenter Name="CP" Foreground="{TemplateBinding Foreground}" Margin="{TemplateBinding Padding}" HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}" VerticalAlignment="{TemplateBinding VerticalContentAlignment}" Content="{TemplateBinding Content}">
				  </ContentPresenter>
			  </Border>			  
		  </ControlTemplate>
	  </Setter>
  </Style>
	<Style Selector="controls|SButton /template/ ContentPresenter>TextBlock">
		<Setter Property="TextDecorations" Value="{Binding $parent[controls:SButton].TextDecorations}" />
	</Style>
</Styles>
```

When the program ran, it reports error:

> System.ArgumentException: 'Unable to resolve type controls:SButton from any of the following locations: '

It runs well if I delete the code:
```
<Style Selector="controls|SButton /template/ ContentPresenter>TextBlock">
		<Setter Property="TextDecorations" Value="{Binding $parent[controls:SButton].TextDecorations}" />
	</Style>
```

It seems the error relates to cannot find its controls:SButton parent. How can I solve this?

**Comment by timunie** [comment, marked as answer]
$parent syntax will search for logical parent, however in your case it's more a visual parent. So try 
`{Binding TextDecorations, RelativeSoruce={RelativeSource Mode=FindAnchestor, AnchestorType=MyButtonType}}`

https://docs.avaloniaui.net/docs/next/guides/data-binding/binding-to-controls#binding-to-an-ancestor
Mind the last warning box: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/fc3a474a-b524-46d1-8461-b28ac21b1c46)

**Reply by kyuranger** [reply]
It works. Thank you!

**Reply by kyuranger** [reply]
HI, I have a new problem. After I bind it as you said, it runs without any error while the output records the error as below:

> [Binding]Error in binding to 'Avalonia.Controls.TextBlock'.'TextDecorations': 'Null value in expression '{empty}' at ''.'(TextBlock #53180767)

Even I have set the default value in XAML and code-behind already:
`<Setter Property="controls:SButton.TextDecorations" Value="Baseline"></Setter>`

```
   public static readonly StyledProperty<TextDecorationCollection?> TextDecorationsProperty =
AvaloniaProperty.Register<Button, TextDecorationCollection?>(nameof(TextDecorations),TextDecorationCollection.Parse("Baseline"));
        public TextDecorationCollection? TextDecorations
        {
            get { return GetValue(TextDecorationsProperty); }
            set { SetValue(TextDecorationsProperty, value); }
        }
```

What's wrong with it?

**Reply by timunie** [reply]
you register the property for Button, not MyButton. So it cannot find the property in MyButton.

**Reply by kyuranger** [reply]
Yes, it works now. Thank you.

**Reply by martinrhan** [reply]
Why is is Anchestor not Ancestor

**Reply by rabbitism** [reply]
> Why is is Anchestor not Ancestor

Just typo.
